### PR TITLE
Add OpenStack provider

### PR DESCRIPTION
**Pull request by tfussell**
For https://github.com/giantswarm/giantswarm/issues/19400

Adds support for templating OpenStack clusters.

Adds the following OpenStack-specific flags:
- `--cloud`
- `--cloud-config`
- `--dns-nameservers`
- `--external-network-id`
- `--failure-domain`
- `--image-name`
- `--node-machine-flavor`
- `--node-cidr`
- `--root-volume-disk-size`
- `--root-volume-source-type`
- `--root-volume-source-uuid`

Some of these flags will be dropped in favor of defaulting. Others will become flags common for all providers. For now we need the configurability for testing OpenStack alpha releases.

### CODE REVIEW

File: cmd/template/cluster/provider/templates/openstack/open_stack_machine_template.yaml.tmpl
Comment (ericgraf) [inline cmd/template/cluster/provider/templates/openstack/open_stack_machine_template.yaml.tmpl:16]: Pass in as a variable. OpenStack project specific. This is the key stored in OpenStack itself.

File: cmd/template/cluster/provider/templates/openstack/open_stack_machine_template.yaml.tmpl
Comment (ericgraf) [inline cmd/template/cluster/provider/templates/openstack/open_stack_machine_template.yaml.tmpl:17]: Very likely same variable as above cr.  Please pass in .
Default likely will be `openstack`.

File: cmd/template/cluster/provider/templates/openstack/open_stack_machine_template.yaml.tmpl
Comment (ericgraf) [inline cmd/template/cluster/provider/templates/openstack/open_stack_machine_template.yaml.tmpl:19]: Very likely same variable as above cr. Please pass in .
Default likely will be `cloud-config`.

File: cmd/template/cluster/provider/templates/openstack/kubeadm_control_plane.yaml.tmpl
Comment (ericgraf) [inline cmd/template/cluster/provider/templates/openstack/kubeadm_control_plane.yaml.tmpl:33]: Why do we have to update the hosts file here ?

File: cmd/template/cluster/provider/templates/openstack/kubeadm_control_plane.yaml.tmpl
Comment (ericgraf) [inline cmd/template/cluster/provider/templates/openstack/kubeadm_control_plane.yaml.tmpl:37]: Cloudinit populates the hostname of the machine. Is there a reason why we had to set this ?

**Review by ericgraf — CHANGES_REQUESTED** [review]

File: cmd/template/cluster/provider/capo.go
Comment (tfussell) [reply cmd/template/cluster/provider/capo.go:25]: Will drop.

File: cmd/template/cluster/provider/templates/openstack/open_stack_machine_template.yaml.tmpl
Comment (yulianedyalkova) [inline cmd/template/cluster/provider/templates/openstack/open_stack_machine_template.yaml.tmpl:14]: Is it `.NodeMachineFlavor` or `.ControlPlaneMachineFlavor` that needs to be set here?

Comment (ericgraf) [reply cmd/template/cluster/provider/templates/openstack/open_stack_machine_template.yaml.tmpl:14]: This template is shared for both machine_deployment.yaml.tmpl and kubeadm_control_plane.yaml.tmpl . Does it make sense to have a control plane machine template that is separate to workers?

There are a number of properties that I could see being different down the road.
https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/main/config/crd/bases/infrastructure.cluster.x-k8s.io_openstackmachinetemplates.yaml#L73

Comment (tfussell) [reply cmd/template/cluster/provider/templates/openstack/open_stack_machine_template.yaml.tmpl:14]: Good point. I want to keep this as simple as possible for now so I will remove `.ControlPlaneMachineFlavor` for now. When we need it we can add a second machine template for the control plane.

File: cmd/template/cluster/provider/templates/openstack/open_stack_cluster.yaml.tmpl
Comment (tfussell) [reply cmd/template/cluster/provider/templates/openstack/open_stack_cluster.yaml.tmpl]: Done.

File: cmd/template/cluster/flag.go
Comment (tfussell) [inline cmd/template/cluster/flag.go:250]: @giantswarm/team-rocket Should we add more validation here?

File: cmd/template/cluster/provider/templates/aws/cluster.yaml.tmpl
Comment (tfussell) [inline cmd/template/cluster/provider/templates/aws/cluster.yaml.tmpl:15]: Made this consistent across providers.

File: cmd/template/cluster/runner.go
Comment (tfussell) [inline cmd/template/cluster/runner.go:100]: Only legacy AWS clusters use the `default` namespace now.

**Review by ericgraf — APPROVED** [review]
LGTM!

**Review by fiunchinho — APPROVED** [review]

**Comment by marians** [comment]
`kubectl gs template cluster --help` is already pretty messy, because of a lot of options, and I am worried that adding another 12 options will render it useless. That's a big side effect for all users, while the flags to be added will only serve a few.

What could we do?

Can me make the OpenStack related flags hidden, as they are meant to be used by a limited audience only (and may be removed in the future anyway)? This way we wouldn't have negative side effects for people using AWS or Azure only. But that would make the flags harder to use for those few who actually need them.

Can we structure the help output in a way that OpenStack-related flags form their own group? That would actually be a big improvement for every user.

**Comment by tfussell** [comment]
@marians I don't really like the idea of hiding them because I want OpenStack customers to be supported just like AWS and Azure customers.

One option would be to add a subcommand under `opsctl template cluster` for each provider. Then, `opsctl template cluster openstack` could only show the flags for that provider.

Another option would be to parse the `--provider` flag immediately then initialize the other flags depending on its value. `opsctl template cluster --provider openstack --help` would show all flags including OpenStack only while `opsctl template cluster --help` would only show common flags. This isn't very intuitive though.

We could also write our own help message instead of using the automatically generated message.

One last option is to keep it like this until `cobra` merges the PR adding flag grouping support. It sounds like it will be soon. I actually saw your comment there :)

What do you think @giantswarm/sig-ux?

**Comment by marians** [comment]
For reference, here is the upstream issue: https://github.com/spf13/cobra/issues/1327

I can't find a PR to close that issue. The linked PR 1003 implements grouping of commands, not flags.

I would suggest to override the flags help rendering then with our own "hard-coded" string.

File: cmd/template/cluster/provider/common.go
Comment (MarcelMue) [inline cmd/template/cluster/provider/common.go:31]: Are all of these options really unique per workload cluster? 
Wouldn't it make way more sense to default a bunch of these (e.g. SSHKeyName, ImageName, ...) in the webhook as they are tied to the management cluster?

Additionally we know that customers want to use gitops to manage these clusters, supplying them with good overlays seems more sensible to me.

File: cmd/template/cluster/provider/templates/openstack/machine_deployment.yaml.tmpl
Comment (MarcelMue) [inline cmd/template/cluster/provider/templates/openstack/machine_deployment.yaml.tmpl:2]: Can we use machinepools on Openstack? Machinedeployments are not nice to manage with gitops.

File: cmd/template/cluster/flag.go
Comment (marians) [inline cmd/template/cluster/flag.go:116]: Are any of these flags required with `--provider openstack`? If yes, it would be great if we could mention that here (e. g. instead of `(OpenStack only)` add `Required with provider `openstack`.

**Comment by marians** [comment]
Regarding the general UX concerns, we just discussed that in SIG UX. That should not block this PR in any way. I created this issue to tackle the necessary improvements independently:

- https://github.com/giantswarm/roadmap/issues/605

**Comment by tfussell** [comment]
> I would suggest to override the flags help rendering then with our own "hard-coded" string.

After a team discussion, we decided to simply hide all of the provider-specific flags until we have a better solution.

File: cmd/template/cluster/provider/common.go
Comment (tfussell) [reply cmd/template/cluster/provider/common.go:31]: I dropped SSH key name because we don't want to allow it. Image name will be decided by us (we're managing images), but is dependent on Kubernetes version and may be useful to override. Source UUID is tricky because it depends on the OpenStack project the cluster is deployed into and that is therefore cluster-dependent.

For the alpha (or "pre-alpha") we just want the ability to specify all of the options. The flags won't show up in the automatic CLI help so I don't think it will be a problem in terms of backwards compatibility. Let me know if you disagree @MarcelMue.

File: cmd/template/cluster/provider/templates/openstack/machine_deployment.yaml.tmpl
Comment (tfussell) [reply cmd/template/cluster/provider/templates/openstack/machine_deployment.yaml.tmpl:2]: We discussed this in Slack.

**Comment by MarcelMue** [comment]
^ sorry for the wrong commit - got mixed up with my CLI.

**Pull request merged**
